### DISCUSSION TITLE
CORS issue while redirecting to Keycloak Login Page

### DISCUSSION THREAD

**Discussion by suryadip08** [keycloak/keycloak#15962] [Q&A]
I have a simple Rest API written in Spring Boot. I have implemented OAuth2 with Keycloak. I have a UI written in Angular.

My expectation is, when I hit the API url from the Angular Page (say a Button click), I should be redirected to Keycloak Login Page. But when I try to access the API from Angular Application, I am getting the below error.

 Access to XMLHttpRequest at 'http://localhost:8080/realms/CloudBrokerRealm/protocol/openid-connect/auth?response_type=code&client_id=cloud-broker-client2&scope=openid&state=pgUtz5IX5ONC3NQ6o9v3IbPnB2DB9rUr3O40tuRxGHY%3D&redirect_uri=http://localhost:8085/login/oauth2/code/keycloak&nonce=CjpTbFvhBqxD0Rz4SQk49rT-wsB3C279NlGHDExlh6M' (**redirected from** **'http://localhost:8085/scheduler/read'**) from origin 'http://localhost:4200' **has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.**

When I hit the API http://localhost:8085/scheduler/read by clicking a Button from Angular, It is trying to redirect to Keycloak Login page but due to CORS error, Login page of Keycloak is not coming.

For accessing the backend API http://localhost:8085/scheduler/read , I have done the CORS configuration in Spring Boot and no CORS issues with this API when I directly accessing this API url from browser and Keycloak Login Page is coming. But the issue is occurring when from Angular Application running in http://localhost:4200 , we are trying to hit the API and it is trying to redirect to Keycloak Login Page http://localhost:8080/realms/CloudBrokerRealm/protoco.......

I have added Web Origins in Keycloak under the client but still CORS issue persists. Is there any way to bypass the CORS issue or can we disable the CORS in Keycloak. I am running Keycloak as a docker container image `quay.io/keycloak/keycloak:18.0.2`.

![8FIqA](https://user-images.githubusercontent.com/18580951/207080702-d32afacd-d11a-4bfc-a06d-79bc93d1b25d.png)

**Comment by SunDrop1** [comment]
i have literally the same problem. Yet no solution found

**Comment by olivierdepriester** [comment]
Afaic I implemented this by having the Angular frontend triggers the authentication by itself through a public clientId. The backend uses a bearer-only clientId.
Thus, the referer and redirect URI are both http://localhost:4200 and I don't meet any CORS issue when the response comes back to the browser. Then the backend APIs authenticate the access token via the bearer-only client. It doesn't care of the CORS anymore.

Actually I set "*" for Web Origins and Valid Redirect URIs of my public clientId but I guess I could have been more restrictive.

**Comment by darius-m** [comment]
Admittedly, I am not familiar with Angular, but as far as I understand, CORS errors occur when the frontend Javascript engine attempts to fetch the response itself from Keycloak using an XMLHttpRequest (which is transparent for the user). However, the SSO needs to validate the request entirely, in a client-agnostic way, and the request cannot be transparent from the user (especially if user consent is required). Because of this, the browser must create an entirely new request to the SSO, and the request should be visible to the user.

From what I can see, a functional approach is creating a `303` or `302` (not sure if one of them is better or required) redirect, so the browser creates an entirely new GET request to Keycloak. If you look at the URL in the XMLHttpRequest, you can see that it contains all the required information for Keycloak - client ID, scope and a redirect URI; after Keycloak validates the client's ID and authenticates the user, it will redirect the user through the browser using another 302/303 redirect with an authentication code. The code (implicit) OAuth2 grant allows accessing user ID information only from the backend - the browser only receives a code that it must pass to the backend, which in turn can use the code and client authentication information to fetch the user ID token from the SSO's token endpoint.

**Comment by amitmahajan82** [comment]
> I have a simple Rest API written in Spring Boot. I have implemented OAuth2 with Keycloak. I have a UI written in Angular.
> 
> My expectation is, when I hit the API url from the Angular Page (say a Button click), I should be redirected to Keycloak Login Page. But when I try to access the API from Angular Application, I am getting the below error.
> 
> Access to XMLHttpRequest at 'http://localhost:8080/realms/CloudBrokerRealm/protocol/openid-connect/auth?response_type=code&client_id=cloud-broker-client2&scope=openid&state=pgUtz5IX5ONC3NQ6o9v3IbPnB2DB9rUr3O40tuRxGHY%3D&redirect_uri=http://localhost:8085/login/oauth2/code/keycloak&nonce=CjpTbFvhBqxD0Rz4SQk49rT-wsB3C279NlGHDExlh6M' (**redirected from** **'http://localhost:8085/scheduler/read'**) from origin 'http://localhost:4200' **has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.**
> 
> When I hit the API http://localhost:8085/scheduler/read by clicking a Button from Angular, It is trying to redirect to Keycloak Login page but due to CORS error, Login page of Keycloak is not coming.
> 
> For accessing the backend API http://localhost:8085/scheduler/read , I have done the CORS configuration in Spring Boot and no CORS issues with this API when I directly accessing this API url from browser and Keycloak Login Page is coming. But the issue is occurring when from Angular Application running in http://localhost:4200 , we are trying to hit the API and it is trying to redirect to Keycloak Login Page http://localhost:8080/realms/CloudBrokerRealm/protoco.......
> 
> I have added Web Origins in Keycloak under the client but still CORS issue persists. Is there any way to bypass the CORS issue or can we disable the CORS in Keycloak. I am running Keycloak as a docker container image `quay.io/keycloak/keycloak:18.0.2`.
> 
> ![8FIqA](https://user-images.githubusercontent.com/18580951/207080702-d32afacd-d11a-4bfc-a06d-79bc93d1b25d.png)

Hi, I am also facing this problem. Did you get a solution for this?

**Comment by StephanyBatista** [comment]
I have the same problem

Access to XMLHttpRequest at 'http://localhost:8080/auth/realms/arqgrifo/.well-known/openid-configuration' from origin 'https://localhost:44337' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

and in my client app

![image](https://user-images.githubusercontent.com/6475229/231200714-4c119c58-efd7-400b-8468-3bf43d08492f.png)
![image](https://user-images.githubusercontent.com/6475229/231200799-6965533d-2e90-4c93-a03b-cbc5d8f29955.png)

Someone can help me?

**Reply by hamza200222** [reply]
if your keycloak is  working fine with spring boot create the rest api that would do the thing that you wanted to do directly with the keycloak here's an example : 
```` 
@RestController
@RequestMapping("/api")
public class TokenController {

    @Autowired
    private RestTemplate restTemplate;

    @GetMapping("/check-token")
    public boolean checkToken(@RequestHeader("Authorization") String token) {
        // Remove "Bearer " prefix if present
        if (token.startsWith("Bearer ")) {
            token = token.substring(7);
        }

        // Prepare the request to the introspection endpoint
        String introspectionUrl = "http://localhost:9098/realms/<yourRealm>/protocol/openid-connect/token/introspect";

        HttpHeaders headers = new HttpHeaders();
        headers.setBasicAuth("<client-id>", "<client-secret>"); // Basic auth for client credentials
        headers.setContentType(org.springframework.http.MediaType.APPLICATION_FORM_URLENCODED);

        MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
        body.add("token", token);

        HttpEntity<MultiValueMap<String, String>> entity = new HttpEntity<>(body, headers);

        // Send the request
        ResponseEntity<Map> response = restTemplate.exchange(
                introspectionUrl, HttpMethod.POST, entity, Map.class
        );

        // Check the response
        Map<String, Object> responseBody = response.getBody();
        if (responseBody != null && Boolean.TRUE.equals(responseBody.get("active"))) {
            return true;
        } else {
            return false;
        }
    }
}

````

**Comment by diegoot-dev** [comment]
**any news?, I have same problem**

**Comment by MaxiCattaneoCvetic** [comment]
Hello, I have the same issue. Im using react and spring cloud gateway for redirect all request to my microservices.
Any news?

**Reply by hamza200222** [reply]
me to I had the same issue try this it might work ⬆⬆⬆

**Comment by alex27riva** [comment]
I have the same issue. I'm using Flutter Web.

**Comment by MaymoonaAlBoloshi** [comment]
same issue right, configs in keykloak re similar, not sure how to solve this

**Reply by Arnaud-Lyard** [reply]
Same issue on NodeJS. How to solve this ?

**Comment by septianrezaandrianto** [comment]
Same issue on React JS right now,
I try to upgrade keycloak from 23.0.4 version to be 26.1.1 version but got Cors origin

**Comment by lxbachit03** [comment]
Dear everyone, i just found solution. **(No CORS problem)**
After reading this [issue](https://keycloak.discourse.group/t/cors-issue-with-keycloak-using-it-in-react-and-web-origins-set/13529), so maybe the pattern e.g. **"*" and "+"** may not work locally. They only seem to work when using a **reverse proxy**.

The solution i found from this [stackoverflow](https://stackoverflow.com/questions/10883211/why-does-my-http-localhost-cors-origin-not-work). 
**npm install -g local-cors-proxy** follow it's instructions and **Keycloak** should work locally.

For me, i integrate Keyclock with NextJS and i found some useful article about 2 stacks technologies with **Next-auth:** : [medium article](https://medium.com/inspiredbrilliance/implementing-authentication-in-next-js-v13-application-with-keycloak-part-1-f4817c53c7ef).

![Screenshot 2025-02-20 104543](https://github.com/user-attachments/assets/54afb81a-bc21-4969-a742-48c9efb5e9ba)
![Screenshot 2025-02-20 104519](https://github.com/user-attachments/assets/79e63fa2-5c8a-4e44-8be2-6f6050664c77)
![Screenshot 2025-02-20 104917](https://github.com/user-attachments/assets/cdbe7aa4-33e9-4c8e-b4d0-c08f5b7c0f3d)


package **local-cors-proxy** worked like a charm for me.

**Reply by annagthb** [reply]
Hi!

I have also faced the same issue and came across this response: https://keycloak.discourse.group/t/cors-error-with-openid-1-0-connect-logout/14139. In summary if you're hitting CORS issues due to $.ajax() or fetch(), the cleanest solution is to use standard form submits or browser redirects when possible — especially for login or logout flows.


Action                             |                      XHR/Fetch Used?            |                        CORS Affected?
`$.ajax(), fetch()   `                                      yes                                                             yes
`window.location.href = `                          no                                                              no
`<form method="POST"> `                       no                                                              no
`<a href="...">  `                                         no                                                              no

However I do not want to stop using my ajax call and XHR, so any other solution would be greatly appreciated!
Many thanks,

**Comment by Dunge** [comment]
I hit this problem while hosting Keycloak on Kubernetes behind an ingress-nginx proxy. I needed to add this annotation to the ingress:

>     nginx.ingress.kubernetes.io/enable-cors: "true"
